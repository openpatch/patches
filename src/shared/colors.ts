export interface Color {
  50: string;
  100: string;
  200: string;
  300: string;
  400: string;
  500: string;
  600: string;
  700: string;
  800: string;
  900: string;
}

export const teal: Color = {
  50: "#EFFCF6",
  100: "#C6F7E2",
  200: "#8EEDC7",
  300: "#65D6AD",
  400: "#3EBD93",
  500: "#27AB83",
  600: "#199473",
  700: "#147D64",
  800: "#0C6B58",
  900: "#014D40",
};

export const blueGrey: Color = {
  50: "#F0F4F8",
  100: "#D9E2EC",
  200: "#BCCCDC",
  300: "#9FB3C8",
  400: "#829AB1",
  500: "#627D98",
  600: "#486581",
  700: "#334E68",
  800: "#243B53",
  900: "#102A43",
};

export const blue: Color = {
  50: "#DCEEFB",
  100: "#B6E0FE",
  200: "#84C5F4",
  300: "#62B0E8",
  400: "#4098D7",
  500: "#2680C2",
  600: "#186FAF",
  700: "#0F609B",
  800: "#0A558C",
  900: "#003E6B",
};

export const purple: Color = {
  50: "#EAE2F8",
  100: "#CFBCF2",
  200: "#A081D9",
  300: "#8662C7",
  400: "#724BB7",
  500: "#653CAD",
  600: "#51279B",
  700: "#421987",
  800: "#34126F",
  900: "#240754",
};

export const red: Color = {
  50: "#FFEEEE",
  100: "#FACDCD",
  200: "#F29B9B",
  300: "#E66A6A",
  400: "#D64545",
  500: "#BA2525",
  600: "#A61B1B",
  700: "#911111",
  800: "#780A0A",
  900: "#610404",
};

export const yellow: Color = {
  50: "#FFFAEB",
  100: "#FCEFC7",
  200: "#F8E3A3",
  300: "#F9DA8B",
  400: "#F7D070",
  500: "#E9B949",
  600: "#C99A2E",
  700: "#A27C1A",
  800: "#7C5E10",
  900: "#513C06",
};
