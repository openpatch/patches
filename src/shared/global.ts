import { createGlobalStyle, css } from "styled-components";

export const fontUrl =
  "https://fonts.googleapis.com/css?family=Nunito+Sans:400,700,800,900";

export const bodyStyles = css`
  font-family: ${({ theme }) => theme.fontFamily.body};
  font-size: ${({ theme }) => theme.fontFamily[1]};
  color: ${({ theme }) => theme.palette.neutrals[900]};

  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
  -webkit-tap-highlight-color: transparent;
  -webkit-overflow-scrolling: touch;

  * {
    box-sizing: border-box;
  }

  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    font-weight: ${({ theme }) => theme.fontWeight[2]};
    margin: 0;
    padding: 0;
  }

  button,
  input,
  textarea,
  select {
    outline: none;
    font-family: ${({ theme }) => theme.fontFamily.body};
  }

  sub,
  sup {
    font-size: 0.8em;
  }

  sub {
    bottom: -0.2em;
  }

  sup {
    top: -0.2em;
  }

  b,
  em {
    font-weight: ${({ theme }) => theme.fontWeight[2]};
  }

  hr {
    border: none;
    border-top: 1px solid;
    border-color: ${({ theme }) => theme.palette.neutrals[300]};
    clear: both;
    margin-bottom: 1.25rem;
  }

  code,
  pre {
    font-family: ${({ theme }) => theme.fontFamily.code};
    font-size: ${({ theme }) => theme.fontSize[1]};
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
  }

  code {
    display: inline-block;
    padding-left: 2px;
    padding-right: 2px;
    vertical-align: baseline;

    color: ${({ theme }) => theme.palette.neutrals[500]};
  }

  pre {
    line-height: 18px;
    padding: 11px 1rem;
    white-space: pre-wrap;

    background: rgba(0, 0, 0, 0.05);
    color: ${({ theme }) => theme.palette.neutrals[900]};
    border-radius: 3px;
    margin: 1rem 0;
  }

  &.ReactModal__Body--open {
    overflow: hidden;
    &.hide-intercom #intercom-container {
      display: none;
    }
  }

  .ReactModalPortal > div {
    opacity: 0;
  }

  .ReactModalPortal .ReactModal__Overlay {
    transition: all 200ms ease-in;

    &--after-open {
      opacity: 1;
    }
    &--before-close {
      opacity: 0;
    }
  }
`;

export const GlobalStyle = createGlobalStyle`
 body {
   ${bodyStyles}
 }
`;
