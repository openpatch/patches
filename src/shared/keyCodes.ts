import { KeyboardEvent } from "react";

export const isEnter = (event: KeyboardEvent): boolean =>
  "key" in event && event.key === "Enter";

export const isSpacebar = (event: KeyboardEvent): boolean =>
  "key" in event && event.key === " ";

export const isArrowLeft = (event: KeyboardEvent): boolean =>
  "key" in event && event.key === "ArrowLeft";

export const isArrowRight = (event: KeyboardEvent): boolean =>
  "key" in event && event.key === "ArrowRight";

export const isArrowDown = (event: KeyboardEvent): boolean =>
  "key" in event && event.key === "ArrowDown";
