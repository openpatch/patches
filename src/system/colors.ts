import { system } from "styled-system";
import { SystemColor, ResponsiveProp } from "./types";

export type ColorProps = {
  textColor?: ResponsiveProp<SystemColor>;
  backgroundColor?: ResponsiveProp<SystemColor>;
  bg?: ResponsiveProp<SystemColor>;
};

export const color = system({
  textColor: {
    scale: "colors",
    property: "color",
  },
  backgroundColor: {
    scale: "colors",
    property: "backgroundColor",
  },
  bg: {
    scale: "colors",
    property: "backgroundColor",
  },
});
