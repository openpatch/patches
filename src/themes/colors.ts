import * as CSS from "csstype";

export type Color = {
  "50": Exclude<CSS.Properties["color"], undefined>;
  "100": Exclude<CSS.Properties["color"], undefined>;
  "200": Exclude<CSS.Properties["color"], undefined>;
  "300": Exclude<CSS.Properties["color"], undefined>;
  "400": Exclude<CSS.Properties["color"], undefined>;
  "500": Exclude<CSS.Properties["color"], undefined>;
  "600": Exclude<CSS.Properties["color"], undefined>;
  "700": Exclude<CSS.Properties["color"], undefined>;
  "800": Exclude<CSS.Properties["color"], undefined>;
  "900": Exclude<CSS.Properties["color"], undefined>;
};

export const teal: Color = {
  "50": "#EFFCF6",
  "100": "#C6F7E2",
  "200": "#8EEDC7",
  "300": "#65D6AD",
  "400": "#3EBD93",
  "500": "#27AB83",
  "600": "#199473",
  "700": "#147D64",
  "800": "#0C6B58",
  "900": "#014D40",
};

export const mintGreen: Color = {
  "50": "#eaffea",
  "100": "#d6ffd6",
  "200": "#ccffcc",
  "300": "#b7ffb7",
  "400": "#adffad",
  "500": "#98ff98",
  "600": "#89e689",
  "700": "#7acc7a",
  "800": "#5b995b",
  "900": "#3d663d",
};

export const bahamaBlue: Color = {
  "50": "#e6f1f4",
  "100": "#b3d4df",
  "200": "#99c5d5",
  "300": "#66a9bf",
  "400": "#008ebd",
  "500": "#006f95",
  "600": "#006282",
  "700": "#005977",
  "800": "#004359",
  "900": "#00384b",
};

export const blueGrey: Color = {
  "50": "#F0F4F8",
  "100": "#D9E2EC",
  "200": "#BCCCDC",
  "300": "#9FB3C8",
  "400": "#829AB1",
  "500": "#627D98",
  "600": "#486581",
  "700": "#334E68",
  "800": "#243B53",
  "900": "#102A43",
};

export const coolGrey: Color = {
  "50": "#F5F7FA",
  "100": "#E4E7EB",
  "200": "#CBD2D9",
  "300": "#9AA5B1",
  "400": "#7B8794",
  "500": "#616E7C",
  "600": "#52606D",
  "700": "#3E4C59",
  "800": "#323F4B",
  "900": "#1F2933",
};

export const blue: Color = {
  "50": "#DCEEFB",
  "100": "#B6E0FE",
  "200": "#84C5F4",
  "300": "#62B0E8",
  "400": "#4098D7",
  "500": "#2680C2",
  "600": "#186FAF",
  "700": "#0F609B",
  "800": "#0A558C",
  "900": "#003E6B",
};

export const blueVivid: Color = {
  "50": "#E6F6FF",
  "100": "#BAE3FF",
  "200": "#7CC4FA",
  "300": "#47A3F3",
  "400": "#2186EB",
  "500": "#0967D2",
  "600": "#0552B5",
  "700": "#03449E",
  "800": "#01337D",
  "900": "#002159",
};

export const cyanVivid: Color = {
  "50": "#E1FCF8",
  "100": "#C1FEF6",
  "200": "#92FDF2",
  "300": "#62F4EB",
  "400": "#3AE7E1",
  "500": "#1CD4D4",
  "600": "#0FB5BA",
  "700": "#099AA4",
  "800": "#07818F",
  "900": "#05606E",
};

export const pinkVivid: Color = {
  "50": "#FFE3EC",
  "100": "#FFB8D2",
  "200": "#FF8CBA",
  "300": "#F364A2",
  "400": "#E8368F",
  "500": "#DA127D",
  "600": "#BC0A6F",
  "700": "#A30664",
  "800": "#870557",
  "900": "#620042",
};

export const purple: Color = {
  "50": "#EAE2F8",
  "100": "#CFBCF2",
  "200": "#A081D9",
  "300": "#8662C7",
  "400": "#724BB7",
  "500": "#653CAD",
  "600": "#51279B",
  "700": "#421987",
  "800": "#34126F",
  "900": "#240754",
};

export const purpleVivid: Color = {
  "50": "#F2EBFE",
  "100": "#DAC4FF",
  "200": "#B990FF",
  "300": "#A368FC",
  "400": "#9446ED",
  "500": "#8719E0",
  "600": "#7A0ECC",
  "700": "#690CB0",
  "800": "#580A94",
  "900": "#44056E",
};

export const orangeVivid: Color = {
  "50": "#FFE8D9",
  "100": "#FFD0B5",
  "200": "#FFB088",
  "300": "#FF9466",
  "400": "#F9703E",
  "500": "#F35627",
  "600": "#DE3A11",
  "700": "#C52707",
  "800": "#AD1D07",
  "900": "#841003",
};

export const red: Color = {
  "50": "#FFEEEE",
  "100": "#FACDCD",
  "200": "#F29B9B",
  "300": "#E66A6A",
  "400": "#D64545",
  "500": "#BA2525",
  "600": "#A61B1B",
  "700": "#911111",
  "800": "#780A0A",
  "900": "#610404",
};

export const redVivid: Color = {
  "50": "#FFE3E3",
  "100": "#FFBDBD",
  "200": "#FF9B9B",
  "300": "#F86A6A",
  "400": "#EF4E4E",
  "500": "#E12D39",
  "600": "#CF1124",
  "700": "#AB091E",
  "800": "#8A041A",
  "900": "#610316",
};

export const yellow: Color = {
  "50": "#FFFAEB",
  "100": "#FCEFC7",
  "200": "#F8E3A3",
  "300": "#F9DA8B",
  "400": "#F7D070",
  "500": "#E9B949",
  "600": "#C99A2E",
  "700": "#A27C1A",
  "800": "#7C5E10",
  "900": "#513C06",
};

export const yellowVivid: Color = {
  "50": "#FFFBEA",
  "100": "#FFF3C4",
  "200": "#FCE588",
  "300": "#FADB5F",
  "400": "#F7C948",
  "500": "#F0B429",
  "600": "#DE911D",
  "700": "#CB6E17",
  "800": "#B44D12",
  "900": "#8D2B0B",
};
