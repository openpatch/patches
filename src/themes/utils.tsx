import React from "react";
import { Global, css } from "@emotion/core";
import {
  Color,
  bahamaBlue,
  blueGrey,
  red,
  yellow,
  blue,
  mintGreen,
} from "../shared/colors";

import { Theme } from "./";

export function cssColor(color: Color, type: string): string {
  const css: Array<string> = [];

  Object.keys(color).forEach((key) => {
    css.push(`--color-${type}-${key}: ${color[key]};`);
  });

  return css.join("\n");
}

export interface GlobalCssThemeProps {
  theme: Theme;
}

export const GlobalCssTheme: React.FC<GlobalCssThemeProps> = ({ theme }) => {
  return (
    <Global
      styles={css`
        :root {
          ${cssColor(theme.colors.primary, "primary")}
          ${cssColor(theme.colors.neutrals, "neutrals")}
          ${cssColor(
            theme.colors.error,
            "error"
          )}
          ${cssColor(
            theme.colors.warning,
            "warning"
          )}
          ${cssColor(
            theme.colors.info,
            "info"
          )}
          ${cssColor(theme.colors.success, "success")}
        }
      `}
    ></Global>
  );
};
