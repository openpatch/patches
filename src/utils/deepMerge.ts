interface IIsObject {
  (item: any): boolean;
}

interface IObject {
  [key: string]: any;
}

interface IDeepMerge {
  (target: IObject, ...sources: Array<IObject>): IObject;
}

/**
 * @description Method to check if an item is an object. Date and Function are considered
 * an object, so if you need to exclude those, please update the method accordingly.
 * @param item - The item that needs to be checked
 * @return {Boolean} Whether or not @item is an object
 */
export const isObject: IIsObject = (item: any): boolean => {
  return item === Object(item) && !Array.isArray(item);
};

/**
 * @description Method to perform a deep merge of objects
 * @param {Object} target - The targeted object that needs to be merged with the supplied @sources
 * @param {Array<Object>} sources - The source(s) that will be used to update the @target object
 * @return {Object} The final merged object
 */
const deepMerge: IDeepMerge = (
  target: IObject,
  ...sources: Array<IObject>
): IObject => {
  // return the target if no sources passed
  if (!sources.length) {
    return target;
  }

  const result: IObject = target;

  if (isObject(result)) {
    const len: number = sources.length;

    for (let i = 0; i < len; i += 1) {
      const elm: any = sources[i];

      if (isObject(elm)) {
        for (const key in elm) {
          if (elm.hasOwnProperty(key)) {
            if (isObject(elm[key])) {
              if (!result[key] || !isObject(result[key])) {
                result[key] = {};
              }
              deepMerge(result[key], elm[key]);
            } else {
              if (Array.isArray(result[key]) && Array.isArray(elm[key])) {
                // concatenate the two arrays and remove any duplicate primitive values
                result[key] = Array.from(new Set(result[key].concat(elm[key])));
              } else {
                result[key] = elm[key];
              }
            }
          }
        }
      }
    }
  }

  return result;
};

export default deepMerge;
