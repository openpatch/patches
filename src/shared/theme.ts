import { DefaultTheme } from "styled-components";
import { blue, purple, red, yellow, teal, blueGrey, Color } from "./colors";

export interface Palette {
  /**
   * These are the splashes of color that should appear the most in your UI,
   and are the ones that determine the overall "look" of the site. Use these
   for things like primary actions, links, navigation items, icons, accent
   borders, or text you want to emphasize.
   */
  primary: Color;
  background: string;
  /**
   * These are the colors you will use the most and will make up the majority
   of your UI. Use them for most of your text, backgrounds, and borders, as
   well as for things like secondary buttons and links.
   */
  neutrals: Color;
  /**
   * These colors should be used fairly conservatively throughout your UI to
   avoid overpowering your primary colors. Use them when you need an element
   to stand out, or to reinforce things like error states or positive trends
   with the appropriate semantic color.
   */
  supporting: {
    accent: Color;
    success: Color;
    info: Color;
    warning: Color;
    error: Color;
  };
}

const lightPalette: Palette = {
  primary: teal,
  neutrals: blueGrey,
  background: "#fff",
  supporting: {
    accent: purple,
    info: blue,
    error: red,
    warning: yellow,
    success: teal,
  },
};

const makeTheme = (palette: Palette): DefaultTheme => ({
  palette,
  screens: {
    sm: 640,
    md: 768,
    lg: 1024,
    xl: 1280,
  },
  spacing: {
    1: 4,
    2: 8,
    3: 12,
    4: 16,
    5: 24,
    6: 32,
    7: 48,
    8: 64,
    9: 96,
    10: 128,
  },
  fontFamily: {
    display: ["Nunito Sans", "sans-serif"].join(", "),
    body: ["Nunito Sans", "sans-serif"].join(", "),
    code: [""].join(", "),
  },
  boxShadow: {
    1: "0 1px 3px rgba(0, 0, 0, .12), 0 1px 2px rgba(0, 0, 0, .24)",
    2: "0 3px 6px rgba(0, 0, 0, .16), 0 2px 4px rgba(0, 0, 0, .12)",
    3: "0 10px 20px rgba(0, 0, 0, .15), 0 3px 6px rgba(0, 0, 0, .10)",
    4: "0 15px 25px rgba(0, 0, 0, .15), 0 5px 10px rgba(0, 0, 0, .05)",
    5: "0 20px 40px rgba(0, 0, 0, .2)",
  },
  borderRadius: {
    1: 4,
    2: 8,
    3: 16,
  },
  borderWidth: {
    1: 1,
    2: 2,
    3: 4,
    4: 8,
  },
  fontSize: {
    1: 12,
    2: 14,
    3: 16,
    4: 18,
    5: 20,
    6: 24,
    7: 30,
    8: 36,
    9: 48,
    10: 60,
  },
  fontWeight: {
    1: 400,
    2: 700,
    3: 800,
    4: 900,
  },
});

export default makeTheme(lightPalette);
