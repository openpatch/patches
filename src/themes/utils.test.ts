import { cssColor } from "./utils";
import { teal } from "../shared/colors";

test("cssColor", () => {
  const css = cssColor(teal, "text");

  expect(css).toMatchSnapshot();
});
