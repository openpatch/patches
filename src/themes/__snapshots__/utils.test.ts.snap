// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cssColor 1`] = `
"--color-text-50 = #EFFCF6;
--color-text-100 = #C6F7E2;
--color-text-200 = #8EEDC7;
--color-text-300 = #65D6AD;
--color-text-400 = #3EBD93;
--color-text-500 = #27AB83;
--color-text-600 = #199473;
--color-text-700 = #147D64;
--color-text-800 = #0C6B58;
--color-text-900 = #014D40;"
`;

exports[`varCssColorMap 1`] = `
Object {
  "100": "var(--color-text-100)",
  "200": "var(--color-text-200)",
  "300": "var(--color-text-300)",
  "400": "var(--color-text-400)",
  "50": "var(--color-text-50)",
  "500": "var(--color-text-500)",
  "600": "var(--color-text-600)",
  "700": "var(--color-text-700)",
  "800": "var(--color-text-800)",
  "900": "var(--color-text-900)",
}
`;
